doc/reference/trd-uart.md

```diff
@@ -0,0 +1,527 @@
+so clients that do not use `receive_word` do not have to implement a
+callback.
+
+If the UART returns `Ok(())` to a call to `receive_word`, it MUST make
+a `received_word` callback in the future, when it receives a data word
+or some error occurs. Valid `Err` values of `receive_word` are:
+  - `BUSY`: the UART is busy with an outstanding call to
+    `receive_buffer` or `receive_word`.
+  - `OFF`: the UART is powered down or in a configuration that does
+    not allow UART reception (e.g., it is a USART in SPI mode).
+  - `NOSUPPORT`: `receive_word` operations are not supported.
+  - `FAIL`: some other error.
+
+5 Composite Traits
+===============================
+
+In addition to the 6 basic traits, the UART HIL defines several traits
+that use these basic traits as supertraits. These composite traits allow
+structures to refer to multiple pieces of UART functionality with a
+single reference and ensure that their implementations are coupled.
+
+```rust
+pub trait Uart<'a>: Configure + Configuration + Transmit<'a> + Receive<'a> {}
+pub trait UartData<'a>: Configuration + Transmit<'a> + Receive<'a> {}
+pub trait UartAdvanced<'a>: Configure + Configuration + Transmit<'a> + ReceiveAdvanced<'a> {}
+pub trait Client: Configuration+ ReceiveClient + TransmitClient {}
+```
+
+The HIL provides blanket implementations of these four traits: any
+structure that implements the supertraits of a composite trait will
+automatically implement the composite trait.
+
+6 Capsules
+===============================
+
+The Tock kernel provides two standard capsules for UARTs:
+
+  * `capsules::console::Console` provides a userspace abstraction of a console.
+    It allows userspace to print to and read from a serial port through a
+    system call API.
+  * `capsules::virtual_uart` provides a set of abstractions for virtualizing
+    a single UART into many UARTs.
+
+The structures in `capsules::virtual_uart` allow multiple clients to
+read from and write to a serial port. Write operations are interleaved
+at the granularity of `transmit_buffer` calls: each client's
+`transmit_buffer` call is printed contiguously, but consecutive calls
+to `transmit_buffer` from a single client may have other data inserted
+between them. When a client calls `receive_buffer`, it starts reading
+data from the serial port at that point in time, for the length of its
+request. If multiple clients make `receive_buffer` calls that overlap
+with one another, they each receive copies of the received data.
+
+Suppose, for example, that there are two clients. One of them calls
+`receive_buffer` for 8 bytes. A user starts typing "1234567890"
+at the console. After the third byte, another client calls `receive_buffer`
+for 4 bytes. After the user types "7", the second client will
+receive a `received_buffer` callback with a buffer containing "4567".
+After the user types "8", the first client will receive a callback
+with a buffer containing "12345678".
```

Comment:
```suggestion
with a buffer containing "12345678". If the second client then calls
`receive_buffer` for 1 byte, they will receive "9" (note that it never
sees "8", since that has been consumed by the time this second
receive call was made).
```

I think this is true?